File: src/main/java/com/xiaorui/agentapplicationcreator/controller/UserController.java
package com.xiaorui.agentapplicationcreator.controller;

import cn.hutool.core.bean.BeanUtil;
import cn.hutool.core.collection.CollUtil;
import cn.hutool.core.util.StrUtil;
import com.mybatisflex.core.paginate.Page;
import com.xiaorui.agentapplicationcreator.common.DeleteRequest;
import com.xiaorui.agentapplicationcreator.constant.UserConstant;
import com.xiaorui.agentapplicationcreator.execption.BusinessException;
import com.xiaorui.agentapplicationcreator.execption.ErrorCode;
import com.xiaorui.agentapplicationcreator.execption.ThrowUtil;
import com.xiaorui.agentapplicationcreator.manager.authority.annotation.AuthCheck;
import com.xiaorui.agentapplicationcreator.manager.ratelimit.RateLimit;
import com.xiaorui.agentapplicationcreator.manager.ratelimit.RateLimitTypeEnum;
import com.xiaorui.agentapplicationcreator.model.dto.user.UserChangeEmailRequest;
import com.xiaorui.agentapplicationcreator.model.dto.user.UserChangePasswordRequest;
import com.xiaorui.agentapplicationcreator.model.dto.user.UserLoginRequest;
import com.xiaorui.agentapplicationcreator.model.dto.user.UserQueryRequest;
import com.xiaorui.agentapplicationcreator.model.dto.user.UserRegisterRequest;
import com.xiaorui.agentapplicationcreator.model.dto.user.UserResetPasswordRequest;
import com.xiaorui.agentapplicationcreator.model.dto.user.UserRoleUpdateRequest;
import com.xiaorui.agentapplicationcreator.model.dto.user.UserSendEmailCodeRequest;
import com.xiaorui.agentapplicationcreator.model.dto.user.UserUnbanRequest;
import com.xiaorui.agentapplicationcreator.model.dto.user.UserUpdateInfoRequest;
import com.xiaorui.agentapplicationcreator.model.entity.User;
import com.xiaorui.agentapplicationcreator.model.vo.UserManageStatsVO;
import com.xiaorui.agentapplicationcreator.model.vo.UserVO;
import com.xiaorui.agentapplicationcreator.response.ServerResponseEntity;
import com.xiaorui.agentapplicationcreator.service.UserService;
import com.xiaorui.agentapplicationcreator.util.SecurityUtil;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import jakarta.annotation.Resource;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.beans.BeanUtils;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * @author xiaorui
 */
@RestController
@RequestMapping("/user")
public class UserController {

    private static final int DEFAULT_PAGE_SIZE = 10;
    private static final int MAX_PUBLIC_PAGE_SIZE = 20;
    private static final int MAX_ADMIN_PAGE_SIZE = 100;

    @Resource
    private UserService userService;

    @PostMapping("/register")
    @Operation(summary = "user register", description = "register user by email verification code")
    @Parameter(name = "userRegisterRequest", description = "user register request")
    public ServerResponseEntity<String> userRegister(@RequestBody UserRegisterRequest userRegisterRequest) {
        ThrowUtil.throwIf(userRegisterRequest == null, ErrorCode.PARAMS_ERROR, "request is blank");
        return ServerResponseEntity.success(userService.userRegister(
                userRegisterRequest.getUserEmail(),
                userRegisterRequest.getLoginPassword(),
                userRegisterRequest.getCheckPassword(),
                userRegisterRequest.getEmailVerifyCode()));
    }

    @PostMapping("/login")
    @Operation(summary = "user login", description = "login by email and password")
    @Parameter(name = "userLoginRequest", description = "user login request")
    public ServerResponseEntity<UserVO> userLogin(@RequestBody UserLoginRequest userLoginRequest, HttpServletRequest request) {
        ThrowUtil.throwIf(userLoginRequest == null, ErrorCode.PARAMS_ERROR, "request is blank");
        boolean result = userService.checkPictureVerifyCode(
                userLoginRequest.getVerifyCode(), userLoginRequest.getServerVerifyCode());
        ThrowUtil.throwIf(!result, ErrorCode.PARAMS_ERROR, "verification code is invalid");
        return ServerResponseEntity.success(
                userService.userLogin(userLoginRequest.getUserEmail(), userLoginRequest.getLoginPassword(), request));
    }

    @PostMapping("/sendEmailCode")
    @Operation(summary = "send email code", description = "send email verification code")
    @Parameter(name = "userSendEmailCodeRequest", description = "send email code request")
    @RateLimit(limitType = RateLimitTypeEnum.USER, rate = 3, rateInterval = 60, message = "email requests are too frequent")
    public ServerResponseEntity<Boolean> sendEmailCode(@RequestBody UserSendEmailCodeRequest userSendEmailCodeRequest,
                                                       HttpServletRequest request) {
        ThrowUtil.throwIf(userSendEmailCodeRequest == null, ErrorCode.PARAMS_ERROR, "request is blank");
        userService.sendEmailCode(userSendEmailCodeRequest.getUserEmail(), userSendEmailCodeRequest.getType(), request);
        return ServerResponseEntity.success(true);
    }

    @GetMapping("/getPictureVerifyCode")
    @Operation(summary = "get picture verification code", description = "get picture verification code")
    public ServerResponseEntity<Map<String, String>> getPictureVerifyCode() {
        return ServerResponseEntity.success(userService.getPictureVerifyCode());
    }

    @GetMapping("/getInfo")
    @Operation(summary = "get current user info", description = "get current user info")
    public ServerResponseEntity<UserVO> getUserInfo() {
        return ServerResponseEntity.success(userService.getUserInfo());
    }

    @PostMapping("/logout")
    @Operation(summary = "user logout", description = "logout current user")
    public ServerResponseEntity<Boolean> userLogout(HttpServletRequest request) {
        return ServerResponseEntity.success(userService.userLogout(request));
    }

    @PostMapping("/change/email")
    @Operation(summary = "change user email", description = "change current user email")
    @Parameter(name = "userChangeEmailRequest", description = "change user email request")
    public ServerResponseEntity<Boolean> changUserEmail(@RequestBody UserChangeEmailRequest userChangeEmailRequest) {
        ThrowUtil.throwIf(userChangeEmailRequest == null, ErrorCode.PARAMS_ERROR, "request is blank");
        return ServerResponseEntity.success(
                userService.changeUserEmail(userChangeEmailRequest.getNewEmail(), userChangeEmailRequest.getEmailVerifyCode()));
    }

    @PostMapping("/change/password")
    @Operation(summary = "change user password", description = "change current user password")
    @Parameter(name = "userChangePasswordRequest", description = "change user password request")
    public ServerResponseEntity<Boolean> changeUserPassword(@RequestBody UserChangePasswordRequest userChangePasswordRequest) {
        ThrowUtil.throwIf(userChangePasswordRequest == null, ErrorCode.PARAMS_ERROR, "request is blank");
        return ServerResponseEntity.success(userService.changeUserPassword(
                userChangePasswordRequest.getOldPassword(),
                userChangePasswordRequest.getNewPassword(),
                userChangePasswordRequest.getCheckPassword()));
    }

    @PostMapping("/reset/password")
    @Operation(summary = "reset user password", description = "reset user password")
    @Parameter(name = "userResetPasswordRequest", description = "reset user password request")
    public ServerResponseEntity<Boolean> resetUserPassword(@RequestBody UserResetPasswordRequest userResetPasswordRequest) {
        ThrowUtil.throwIf(userResetPasswordRequest == null, ErrorCode.PARAMS_ERROR, "request is blank");
        return ServerResponseEntity.success(userService.resetUserPassword(
                userResetPasswordRequest.getUserEmail(),
                userResetPasswordRequest.getEmailVerifyCode(),
                userResetPasswordRequest.getNewPassword(),
                userResetPasswordRequest.getCheckPassword()));
    }

    @PostMapping("/update/info")
    @Operation(summary = "update user info", description = "update current user info")
    @Parameter(name = "userUpdateInfoRequest", description = "update user info request")
    public ServerResponseEntity<Boolean> updateUserInfo(@RequestBody UserUpdateInfoRequest userUpdateInfoRequest) {
        ThrowUtil.throwIf(userUpdateInfoRequest == null, ErrorCode.PARAMS_ERROR, "request is blank");
        String userId = SecurityUtil.getUserInfo().getUserId();
        if (!userId.equals(userUpdateInfoRequest.getUserId())) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR, "can only update current user");
        }
        User user = new User();
        BeanUtils.copyProperties(userUpdateInfoRequest, user);
        user.setUpdateTime(LocalDateTime.now());
        return ServerResponseEntity.success(userService.updateById(user));
    }

    @PostMapping("/getInfoList")
    @Operation(summary = "list users", description = "query users by id or nickname")
    @Parameter(name = "userQueryRequest", description = "user query request")
    public ServerResponseEntity<Page<UserVO>> getUserInfoByIdOrName(@RequestBody UserQueryRequest userQueryRequest) {
        ThrowUtil.throwIf(userQueryRequest == null, ErrorCode.PARAMS_ERROR, "request is blank");
        normalizePageRequest(userQueryRequest, MAX_PUBLIC_PAGE_SIZE);
        int current = userQueryRequest.getCurrent();
        int pageSize = userQueryRequest.getPageSize();
        Page<User> userPage = userService.page(new Page<>(current, pageSize), userService.getQueryWrapper(userQueryRequest));
        Page<UserVO> userInfoPage = new Page<>(current, pageSize, userPage.getTotalRow());
        userInfoPage.setRecords(getUserVOList(userPage.getRecords()));
        return ServerResponseEntity.success(userInfoPage);
    }

    @PostMapping("/update/avatar")
    @Operation(summary = "update user avatar", description = "update current user avatar")
    @Parameter(name = "multipartFile", description = "avatar file")
    public ServerResponseEntity<String> updateUserAvatar(@RequestParam("multipartFile") MultipartFile multipartFile) {
        ThrowUtil.throwIf(multipartFile == null, ErrorCode.PARAMS_ERROR, "file is blank");
        return ServerResponseEntity.success(userService.updateUserAvatar(multipartFile));
    }

    @PostMapping("/list/page/info")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    @Operation(summary = "list users by page", description = "admin list users by page")
    @Parameter(name = "userQueryRequest", description = "user query request")
    public ServerResponseEntity<Page<UserVO>> listUserInfoByPage(@RequestBody UserQueryRequest userQueryRequest) {
        ThrowUtil.throwIf(userQueryRequest == null, ErrorCode.PARAMS_ERROR, "request is blank");
        normalizePageRequest(userQueryRequest, MAX_ADMIN_PAGE_SIZE);
        int current = userQueryRequest.getCurrent();
        int pageSize = userQueryRequest.getPageSize();
        Page<User> userPage = userService.page(new Page<>(current, pageSize), userService.getQueryWrapper(userQueryRequest));
        Page<UserVO> userInfoPage = new Page<>(current, pageSize, userPage.getTotalRow());
        userInfoPage.setRecords(userService.getUserInfoList(userPage.getRecords()));
        return ServerResponseEntity.success(userInfoPage);
    }

    @GetMapping("/get/{userId}")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    @Operation(summary = "get user by id", description = "admin get user by id")
    @Parameter(name = "userId", description = "user id")
    public ServerResponseEntity<User> getUserById(@PathVariable String userId) {
        ThrowUtil.throwIf(userId == null, ErrorCode.PARAMS_ERROR);
        User user = userService.getById(userId);
        ThrowUtil.throwIf(user == null, ErrorCode.NOT_FOUND_ERROR);
        return ServerResponseEntity.success(user);
    }

    @PostMapping("/delete")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    @Operation(summary = "delete user", description = "admin delete a user")
    @Parameter(name = "deleteRequest", description = "delete request")
    public ServerResponseEntity<Boolean> deleteUser(@RequestBody DeleteRequest deleteRequest) {
        ThrowUtil.throwIf(deleteRequest == null, ErrorCode.PARAMS_ERROR, "request is blank");
        ThrowUtil.throwIf(StrUtil.isBlank(deleteRequest.getId()), ErrorCode.PARAMS_ERROR, "user id is blank");
        return ServerResponseEntity.success(userService.removeById(deleteRequest.getId()));
    }

    @PostMapping("/delete/batch")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    @Operation(summary = "batch delete users", description = "admin batch delete users")
    @Parameter(name = "deleteRequestList", description = "batch delete request list")
    public ServerResponseEntity<Boolean> deleteBatchUser(@RequestBody List<DeleteRequest> deleteRequestList) {
        ThrowUtil.throwIf(CollUtil.isEmpty(deleteRequestList), ErrorCode.PARAMS_ERROR, "request is blank");
        List<String> idList = deleteRequestList.stream().map(DeleteRequest::getId).collect(Collectors.toList());
        return ServerResponseEntity.success(userService.removeByIds(idList));
    }

    @PostMapping("/ban")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    @Operation(summary = "ban or unban user", description = "admin ban or unban user")
    @Parameter(name = "userUnbanRequest", description = "ban or unban user request")
    public ServerResponseEntity<Boolean> banOrUnbanUser(@RequestBody UserUnbanRequest userUnbanRequest) {
        ThrowUtil.throwIf(userUnbanRequest == null, ErrorCode.PARAMS_ERROR, "request is blank");
        User admin = userService.getById(SecurityUtil.getUserInfo().getUserId());
        return ServerResponseEntity.success(
                userService.banOrUnbanUser(userUnbanRequest.getUserId(), userUnbanRequest.getIsUnban(), admin));
    }

    @PostMapping("/update/role")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    @Operation(summary = "update user role", description = "admin update user role")
    @Parameter(name = "userRoleUpdateRequest", description = "update user role request")
    public ServerResponseEntity<Boolean> updateUserRole(@RequestBody UserRoleUpdateRequest userRoleUpdateRequest) {
        ThrowUtil.throwIf(userRoleUpdateRequest == null, ErrorCode.PARAMS_ERROR, "request is blank");
        User admin = userService.getById(SecurityUtil.getUserInfo().getUserId());
        return ServerResponseEntity.success(
                userService.updateUserRole(userRoleUpdateRequest.getUserId(), userRoleUpdateRequest.getUserRole(), admin));
    }

    @GetMapping("/admin/stats")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    @Operation(summary = "get user manage stats", description = "admin get user management stats")
    public ServerResponseEntity<UserManageStatsVO> getUserManageStats() {
        return ServerResponseEntity.success(userService.getUserManageStats());
    }

    private UserVO getUserVO(User user) {
        if (user == null) {
            return null;
        }
        UserVO userVO = new UserVO();
        BeanUtil.copyProperties(user, userVO);
        return userVO;
    }

    private List<UserVO> getUserVOList(List<User> userList) {
        if (CollUtil.isEmpty(userList)) {
            return new ArrayList<>();
        }
        return userList.stream().map(this::getUserVO).collect(Collectors.toList());
    }

    private void normalizePageRequest(UserQueryRequest userQueryRequest, int maxPageSize) {
        if (userQueryRequest.getCurrent() <= 0) {
            userQueryRequest.setCurrent(1);
        }
        if (userQueryRequest.getPageSize() <= 0) {
            userQueryRequest.setPageSize(DEFAULT_PAGE_SIZE);
            return;
        }
        ThrowUtil.throwIf(userQueryRequest.getPageSize() > maxPageSize, ErrorCode.PARAMS_ERROR, "pageSize exceeds limit");
    }
}

File: app-creator-fronted/app-creator-fronted/src/api/userController.ts
// @ts-ignore
/* eslint-disable */
import request from '@/request'

/** 封禁或解封用户 管理员封禁或解封用户 POST /user/ban */
export async function banOrUnbanUser(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.banOrUnbanUserParams,
  body: API.UserUnbanRequest,
  options?: { [key: string]: any }
) {
  return request<API.ServerResponseEntityBoolean>('/user/ban', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    params: {
      ...params,
    },
    data: body,
    ...(options || {}),
  })
}

/** 修改用户邮箱 修改当前用户邮箱 POST /user/change/email */
export async function changUserEmail(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.changUserEmailParams,
  body: API.UserChangeEmailRequest,
  options?: { [key: string]: any }
) {
  return request<API.ServerResponseEntityBoolean>('/user/change/email', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    params: {
      ...params,
    },
    data: body,
    ...(options || {}),
  })
}

/** 修改用户密码 修改当前用户密码 POST /user/change/password */
export async function changeUserPassword(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.changeUserPasswordParams,
  body: API.UserChangePasswordRequest,
  options?: { [key: string]: any }
) {
  return request<API.ServerResponseEntityBoolean>('/user/change/password', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    params: {
      ...params,
    },
    data: body,
    ...(options || {}),
  })
}

/** 删除用户 管理员删除指定用户 POST /user/delete */
export async function deleteUser(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.deleteUserParams,
  body: API.DeleteRequest,
  options?: { [key: string]: any }
) {
  return request<API.ServerResponseEntityBoolean>('/user/delete', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    params: {
      ...params,
    },
    data: body,
    ...(options || {}),
  })
}

/** 批量删除用户 管理员批量删除用户 POST /user/delete/batch */
export async function deleteBatchUser(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.deleteBatchUserParams,
  body: API.DeleteRequest[],
  options?: { [key: string]: any }
) {
  return request<API.ServerResponseEntityBoolean>('/user/delete/batch', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    params: {
      ...params,
    },
    data: body,
    ...(options || {}),
  })
}

/** 根据 id 获取用户信息 管理员根据 id 获取用户信息 GET /user/get/${param0} */
export async function getUserById(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getUserByIdParams,
  options?: { [key: string]: any }
) {
  const { userId: param0, ...queryParams } = params
  return request<API.ServerResponseEntityUser>(`/user/get/${param0}`, {
    method: 'GET',
    params: { ...queryParams },
    ...(options || {}),
  })
}

/** 获取用户信息 获取当前用户信息 GET /user/getInfo */
export async function getUserInfo(options?: { [key: string]: any }) {
  return request<API.ServerResponseEntityUserVO>('/user/getInfo', {
    method: 'GET',
    ...(options || {}),
  })
}

/** 用户查询 用户根据用户id、昵称，分页获取查询用户信息 POST /user/getInfoList */
export async function getUserInfoByIdOrName(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getUserInfoByIdOrNameParams,
  body: API.UserQueryRequest,
  options?: { [key: string]: any }
) {
  return request<API.ServerResponseEntityPageUserVO>('/user/getInfoList', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    params: {
      ...params,
    },
    data: body,
    ...(options || {}),
  })
}

/** 获取图形验证码 获取图形验证码 GET /user/getPictureVerifyCode */
export async function getPictureVerifyCode(options?: { [key: string]: any }) {
  return request<API.ServerResponseEntityMapStringString>('/user/getPictureVerifyCode', {
    method: 'GET',
    ...(options || {}),
  })
}

/** 分页查询用户信息 管理员分页查询用户信息 POST /user/list/page/info */
export async function listUserInfoByPage(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.listUserInfoByPageParams,
  body: API.UserQueryRequest,
  options?: { [key: string]: any }
) {
  return request<API.ServerResponseEntityPageUserVO>('/user/list/page/info', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    params: {
      ...params,
    },
    data: body,
    ...(options || {}),
  })
}

/** 用户登录 用户使用邮箱 + 密码登录 POST /user/login */
export async function userLogin(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.userLoginParams,
  body: API.UserLoginRequest,
  options?: { [key: string]: any }
) {
  return request<API.ServerResponseEntityUserVO>('/user/login', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    params: {
      ...params,
    },
    data: body,
    ...(options || {}),
  })
}

/** 用户登出 当前用户登出 POST /user/logout */
export async function userLogout(options?: { [key: string]: any }) {
  return request<API.ServerResponseEntityBoolean>('/user/logout', {
    method: 'POST',
    ...(options || {}),
  })
}

/** 用户注册 用户使用邮箱验证码注册 POST /user/register */
export async function userRegister(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.userRegisterParams,
  body: API.UserRegisterRequest,
  options?: { [key: string]: any }
) {
  return request<API.ServerResponseEntityString>('/user/register', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    params: {
      ...params,
    },
    data: body,
    ...(options || {}),
  })
}

/** 重置用户密码 重置当前用户密码 POST /user/reset/password */
export async function resetUserPassword(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.resetUserPasswordParams,
  body: API.UserResetPasswordRequest,
  options?: { [key: string]: any }
) {
  return request<API.ServerResponseEntityBoolean>('/user/reset/password', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    params: {
      ...params,
    },
    data: body,
    ...(options || {}),
  })
}

/** 发送邮箱验证码 向目标邮箱发送验证码 POST /user/sendEmailCode */
export async function sendEmailCode(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.sendEmailCodeParams,
  body: API.UserSendEmailCodeRequest,
  options?: { [key: string]: any }
) {
  return request<API.ServerResponseEntityBoolean>('/user/sendEmailCode', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    params: {
      ...params,
    },
    data: body,
    ...(options || {}),
  })
}

/** 更新用户头像 更新当前用户头像 POST /user/update/avatar */
export async function updateUserAvatar(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.updateUserAvatarParams,
  body: {},
  options?: { [key: string]: any }
) {
  const file = (body as any)?.multipartFile ?? body
  const formData = new FormData()
  formData.append('multipartFile', file)
  return request<API.ServerResponseEntityString>('/user/update/avatar', {
    method: 'POST',
    params: {
      ...params,
    },
    data: formData,
    ...(options || {}),
  })
}

/** 更新用户信息 更新当前用户信息 POST /user/update/info */
export async function updateUserInfo(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.updateUserInfoParams,
  body: API.UserUpdateInfoRequest,
  options?: { [key: string]: any }
) {
  return request<API.ServerResponseEntityBoolean>('/user/update/info', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    params: {
      ...params,
    },
    data: body,
    ...(options || {}),
  })
}
